### PR TITLE
remove max_results limit in Scopus Search API

### PR DESCRIPTION
**Pull request by gilad1bl**
by adding the "cursor" parameter to headers Elsevier allows an iteration over any number of results set. for now this feature is only available in the Scopus Search API. To use it, just add "cursor=True" when calling ScopusSearch. If "cursor=True" is not specified, the original procedure for fetching results is used (start+count).

### CODE REVIEW

**Comment by crew102** [comment]
Might I suggest you not include a parameter for `cursor`,  but rather just assume it should always be used when it's needed? By adding the ability to set the `cursor` param to False, all you're doing is allowing the user to pull down a very small subset of their data, which I kinda doubt many users will want to do. In other words, I think you should just assume that the user want's the entire result set starting with the `start` entry...This is what happens currently in the package - the user can choose where to start downloading the result set, but it's assumed they want all the entries after that `start` value.

**Comment by gilad1bl** [comment]
Hey, @crew102  - not sure I understand. setting the cursor to False just keeps the code exactly as it was before (which is the default behavior, you can also ignore it and nothing will change in the procedure). just as before, the user will get all the results starting from "start".
setting it to True, will replace the start param by the cursor param, and than it will iterate on all of the results.
Are you suggesting setting the default value to cursor=True?

**Comment by crew102** [comment]
I'm suggesting you remove the parameter `cursor` and just always use the cursor when you it's needed. In other words, don't give the user the option to set a `cursor` parameter, as they will always want to use it when it's needed anyway.

**Comment by crew102** [comment]
Oh, OK, I see now that the API won't allow you to use the `start` param if you choose to use a cursor. My initial thought is to deprecate the `start` param in the `Search` class and always use a cursor instead, with the idea being that people are going to very rarely want to only download partial result sets or start downloading result sets in the middle of the set. Thoughts? Also, @Michael-E-Rose, wanna weigh in on this?

File: scopus/classes/search.py [hunk 1]
```diff
@@ -48,6 +48,10 @@ def __init__(self, query, api, refresh, count=200, start=0,
             COMPLETE.
             Note: Only the ScopusSearch API additionally uses view COMPLETE.
 
+        cursor : str (optional, default = False)
+            Using 'cursor' instead of start itterates over all search results with no limit on the number of the resutls.
```
Comment (Michael-E-Rose) [inline scopus/classes/search.py]: typo: It should be "iterates" and "results".

File: scopus/classes/search.py [hunk 2]
```diff
@@ -76,24 +80,45 @@ def __init__(self, query, api, refresh, count=200, start=0,
...
-            self._json = res.get('search-results', {}).get('entry', [])
-            if n == 0:
-                self._json = ""
-            # Download the remaining information in chunks
-            while n > 0:
-                n -= count
-                start += count
-                params.update({'count': count, 'start': start})
+
+            if cursor == False:
```
Comment (Michael-E-Rose) [inline scopus/classes/search.py]: better: `if not cursor:`

File: scopus/classes/search.py [hunk 3]
```diff
@@ -76,24 +80,45 @@ def __init__(self, query, api, refresh, count=200, start=0,
...
+                            'number, change your query ({}) or set cursor=True'.format(n, query))
+                    raise ScopusQueryError(text)
+                self._json = res.get('search-results', {}).get('entry', [])
+                if n == 0:
+                    self._json = ""
+                # Download the remaining information in chunks
+                while n > 0:
+                    n -= count
+                    params.update({'count': count, 'start': start})
+                    print(params)
```
Comment (Michael-E-Rose) [inline scopus/classes/search.py]: I guess this was for testing, wasn't it?

File: scopus/classes/search.py [hunk 4]
```diff
@@ -76,24 +80,45 @@ def __init__(self, query, api, refresh, count=200, start=0,
...
+                    res = download(url=URL[api], params=params, accept="json").json()
+                    self._json.extend(res.get('search-results', {}).get('entry', []))
+                # Finally write out the file
+                with open(qfile, 'wb') as f:
+                    for item in self._json:
+                        f.write('{}\n'.format(dumps(item)).encode('utf-8'))
+            else:
+                params = {'query': query, 'count': count, 'cursor': '*', 'view': view} # see https://dev.elsevier.com/tecdoc_developer_faq.html
```
Comment (Michael-E-Rose) [inline scopus/classes/search.py]: This and line 85 can be combined: Somewhere before the new line 84, create the `params` dict with items used in both cases. Then in line 85 (old), update `params` with the count. Then in this line, update `params` with items used for parsing with the cursor.

File: scopus/classes/search.py [hunk 5]
```diff
@@ -76,24 +80,45 @@ def __init__(self, query, api, refresh, count=200, start=0,
...
+                res = download(url=URL[api], params=params, accept="json").json()
+                n = int(res['search-results'].get('opensearch:totalResults', 0))
+                if n == 0:
+                    self._json = ""
+                self._json = res.get('search-results', {}).get('entry', [])
+                # Download the remaining information in chunks
+                while n > 0:
+                    next = res['search-results']['cursor'].get('@next')
```
Comment (Michael-E-Rose) [inline scopus/classes/search.py]: `next` should be avoided because it is a Python functional.

File: scopus/classes/search.py [hunk 6]
```diff
@@ -76,24 +80,45 @@ def __init__(self, query, api, refresh, count=200, start=0,
...
-                raise ScopusQueryError(text)
-            self._json = res.get('search-results', {}).get('entry', [])
-            if n == 0:
-                self._json = ""
-            # Download the remaining information in chunks
-            while n > 0:
-                n -= count
-                start += count
-                params.update({'count': count, 'start': start})
+
```
Comment (Michael-E-Rose) [inline scopus/classes/search.py]: There should be no empty line here.

File: scopus/classes/search.py [hunk 7]
```diff
@@ -76,24 +80,45 @@ def __init__(self, query, api, refresh, count=200, start=0,
...
+                with open(qfile, 'wb') as f:
+                    for item in self._json:
+                        f.write('{}\n'.format(dumps(item)).encode('utf-8'))
+            else:
+                params = {'query': query, 'count': count, 'cursor': '*', 'view': view} # see https://dev.elsevier.com/tecdoc_developer_faq.html
+                res = download(url=URL[api], params=params, accept="json").json()
+                n = int(res['search-results'].get('opensearch:totalResults', 0))
+                if n == 0:
+                    self._json = ""
+                self._json = res.get('search-results', {}).get('entry', [])
```
Comment (Michael-E-Rose) [inline scopus/classes/search.py]: This doesn't make sense, if you consider the previous two lines.  They should be swapped.

**Comment by Michael-E-Rose** [comment]
The PR is already a good start, thanks for the work @gilad1bl!  (As said in #84 I have troubles understanding the impact of the cursor, but let's discuss that in the issue conversation.)

The provided code here has some minor issues that can easily be addressed.  What concerns me more is that a lot of the code is redundant.  The code for writing out exists twice: It suffices to put this at the end of the function.  Also, the only thing that differs basically is the `params` dict and how the json is parsed.  This can be combined much more.  Instead of having two blocks, keep the original blocks and every now and then have statements defining different initializations depending on whether cursor is True or not.

Finally I think we should make cursor `True` by default for the ScopusSearch class.  Our typical user is a subscriber, and it's for her that we write our code.

Am I right assuming that cursor is currently only implemented for the Scopus search but neither for author nor affiliation search?

**Comment by crew102** [comment]
* > Am I right assuming that cursor is currently only implemented for the Scopus search but neither for author nor affiliation search?

Looks like cursor will work for the affiliation search API as well. I assume the same is true for the author search. 

* > Finally I think we should make cursor True by default for the ScopusSearch class. Our typical user is a subscriber, and it's for her that we write our code.

I think the cursor param will work for non-subscribers as well. In other words, shouldn't they be able to download, say, 1,000 records using the cursor just as a subscriber would? I'm a subscriber so I'm not sure. 

Finally, I don't think that we should actually have a `cursor` param in the classes. I think we should just always just the cursor for the user under the hood and not allow them to choose between iterating through result sets using the `start` approach versus the `cursor` approach. Make sense? There's a bit more commentary on this thought in my earlier posts up above.

**Comment by Michael-E-Rose** [comment]
I should check in the next days how non-subscribers experience the REST API. It is generally desirable to have few parameters, so it makes sense what you say, but we should not break anything.

**Comment by Michael-E-Rose** [comment]
Hi @gilad1bl! Are you able to improve this PR a bit more in the near future?

**Comment by crew102** [comment]
K, I was wrong. Only subscribers are authorized to use the cursor parameter. I lost my subscription and now I get an error saying `Use of the cursor parameter is restricted` when trying to use the cursor. So this means we will need to actually have a `cursor` parameter that users can specify as either True or False, depending on whether they're subscribers or not.

**Pull request merged**